[오목 1, 2단계] 모찌 미션 제출합니다.

wondroid-world opened the pull request:
# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [X] 프로그램이 정상적으로 작동하는가?
- [X] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [X] 코딩 스타일 가이드를 준수하였는가?
    - [X] IDE 코드 자동 정렬을 적용하였는가?
    - [X] 린트 검사를 통과하였는가?
- [X] 프로그래밍 요구 사항을 준수하였는가?
- [ ] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

<!--
학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
-->

-

## 코드 리뷰 커뮤니케이션

<!--
리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
-->

안녕하세요! 제이든,
저는 안드로이드 7기 모찌입니다. 만나서, 반갑습니다.

저는 이번에 TopDown 방식으로 하면서, 각 객체가 어떤 일을 하는 지, 설계를 하고 코드를 작성했습니다.
하지만,  구현을 하고 보니, 설계가 잘못되었고 제가 상태 패턴을 확실히 잘 몰라서, 완벽한 객체가 탄생하지 못했습니다.

이번에 크게 고려했던 점은 2가지 였습니다.
interface로 결합은 느슨하게 만들고 상태패턴을 적용하여, player는 상태만 가지고 있게 하자는 뜻을 가졌습니다. 

하지만, 상태패턴 중 finish에서 lose, win으로 나뉘는 데, 각자 하는 역할이 없이 상태만 들고 있어서, 메서드 없는 객체를 만들어서 아주 불편한 상태가 되었습니다. 그리고 blackPlayerState의 turn이 끝나면, 원래는 whitePlayerState로 넘기고 싶었으나, 실패를 해서, 결국 자기 자신의 상태를 불변객체로 업데이트 하게 만들었습니다. 

저의 부족한 코드 리뷰 잘 부탁드립니다! 
📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

<!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

<!-- 필요하다면 스크린샷을 첨부하세요. -->

![]()

wondroid-world committed:
docs: 오목 기능 목록 명세 작성

wondroid-world committed:
feat: X좌표 객체 구현

- X의 범위는 A~O까지이다.

wondroid-world committed:
feat: Y좌표 객체 구현

- Y좌표의 범위는 1에서 15까지이다.

wondroid-world committed:
feat: Stone interface 구현

- stone 좌표
- stone 색

parkjiminnnn committed:
chore: gitkeep삭제

parkjiminnnn committed:
feat: Stones 객체 구현

parkjiminnnn committed:
feat: x좌표의 알파벳을 숫자로 변경하는 기능 구현

parkjiminnnn committed:
refactor: 함수명 변경

parkjiminnnn committed:
feat: 초기오목판 객체 구현

parkjiminnnn committed:
feat: 오목판에 착수하는 기능 구현

parkjiminnnn committed:
refactor: 인수 변경

parkjiminnnn committed:
chore: 디렉토리 이동

parkjiminnnn committed:
feat: 플레이어의 상태 관련 객체 구현

parkjiminnnn committed:
feat: 플레이어 객체 구현

parkjiminnnn committed:
chore: 디렉토리 이동

parkjiminnnn committed:
feat: view 구현

wondroid-world committed:
feat: 오목 판단 기능 구현

wondroid-world committed:
chore: 디렉토리 이동

wondroid-world committed:
feat: plyaer에 생성자 인자 추가

wondroid-world committed:
test: 오목판 테스트 작성

wondroid-world committed:
feat: turn을 넘기는 기능 구현

wondroid-world committed:
refactor: stone state로 수정

wondroid-world committed:
feat: 게임 실행 구현

wondroid-world committed:
chore: 클래스 삭제

wondroid-world committed:
feat: 렌주룰 기능 구현

wondroid-world committed:
fix: "가 -> 이"로 조사 수정

wondroid-world committed:
refactor: 함수 분리

wondroid-world committed:
feat: 유효성 검사 로직 추가

parkjiminnnn committed:
refactor: 불필요한 로직 제거

parkjiminnnn committed:
style: 주석 제거

parkjiminnnn committed:
refactor: 프로퍼티 제거

parkjiminnnn committed:
feat: 33, 44 로직 구현

parkjiminnnn committed:
feat: 상수 정의

wondroid-world committed:
fix : 소문자를 대문자로 변경

parkjiminnnn committed:
refactor: 함수명 변경

parkjiminnnn committed:
refactor: 프로퍼티 제거

KwonDae commented on src/main/kotlin/omok/model/rule/OmokAdapter.kt:
```diff
@@ -0,0 +1,22 @@
+import omok.model.board.Position
+import omok.model.stone.StoneState
+
+object OmokAdapter {
```
어댑터 패턴을 잘 활용해주셨네요!
현재는 '렌주룰' 하나 이지만, 만약 여러 룰이 적용되어야 한다면 어떻게 개선할 수 있을까요?

wondroid-world replied:
[ 관련 커밋 ] 385803d2a3302765c204db35a09105b3ecf86635

[ 이유 ]
제이든이 이야기한 것처럼 다른 룰이 추가되었을 때에도 사용할 수 있게 rule을 interface로 만들었습니다. 그리고 renjuRule이 Rule을 받아서 구현하도록 하였습니다. 이러면, 다른 룰이 추가되어도 omokGame에 바로 결합할 수 있는 형태가 된다고 생각합니다!

KwonDae commented on src/main/kotlin/omok/model/rule/OmokRule.kt:
```diff
@@ -0,0 +1,98 @@
+    abstract fun validate(
+        board: List<List<Int>>,
+        position: Pair<Int, Int>,
+    ): Boolean
```
OmokRule 추상화를 잘 해주셨어요 👍 
다만, validate를 각 rule 마다 구현을 하고 있는데
지금 만들어주신 3x3, 4x4는 사실 `렌주룰` 이라는 카테고리 안에서 검증해야 할 항목들이 아닐까요?

만약 여기서 렌주룰 외 다른 룰이 추가된다면 어떤 상황이 발생할까요?

wondroid-world replied:
제이든! 제이든 말처럼 렌주룰 안에서 다른 룰이 추가가 된다면, 클래스를 수정을 해야할거 같아요!

제가 위처럼 33, 44, 오목등 각각의 클래스를 만든 이유는 다음과 같습니다. 일단,  playerState에서 33, 44, 오목 등 각각의 조건마다 다른 결과를 반환해야한다고 생각했습니다. 그래서, 각각을 다른 클래스로 구현해서, 각 클래스마다 결과를 보여주고 싶었습니다. 

그리고 renjuRule이 abstract class로 되어있어서, 이를 상속받은 객체가 메시징을 던질 수 있다고 판단했습니다.  

차라리 renjuRule이라는 클래스만들어서 메시징을 던지게 해야하는 걸까요?

KwonDae commented on README.md:
```diff
@@ -1 +1,39 @@
\ No newline at end of file
+# kotlin-omok
+
+## 🛠️ 구현할 기능
```
<img width="453" alt="image" src="https://github.com/user-attachments/assets/a79f3d6f-b90d-467f-a369-fafbb70cd1ba" />

3-3 금수에 돌을 놓으니 'X' 표시와 함꼐 턴이 넘어가고 있네요?
착수금지로 돌을 두지 못하기에 턴이 넘어가는 것은 아닌걸로 알고있는데 다른걸까요?

KwonDae replied:
<img width="913" alt="image" src="https://github.com/user-attachments/assets/f882d45a-1ae7-4c5d-b0db-bae7968ef6de" />

또한 수업자료와의 흑,백이 반대로 되어있는거 같아요. 의도하신걸까요?

wondroid-world replied:
아! 네, console에 찍으니, 빈 원형이 흑같이 보이고 색이 차 있는 원형이 백같이 보여서 조정했습니다! 혹시, 임의로 변경을 한것이 요구 사항을 지키지 않은 것일까요...? 만약 그렇다면 바로 조정하겠습니다!

wondroid-world replied:
네, 콘솔에 찍으니 빈 돌이 흑돌처럼 보이고 원형에 색이 가득찬 돌이 백색으로 보여서 두개를 view에서 바꿔서 구현했습니다!

KwonDae commented on src/main/kotlin/omok/model/player/state/PlayerState.kt:
```diff
@@ -0,0 +1,15 @@
+        stoneState: StoneState,
+    ): PlayerState
+
+    fun nextTurn(): PlayerState
```
PlayerState의 역할이 궁금합니다.
참가자가 보드와 좌표를 받아 스스로 판단하는 흐름이 되는걸까요?

프로그래밍 요구 사항 중 객체지향 원칙의 `단일 책임 원칙` 부터 고민해보면 어떨까요?

> SRP (단일책임의 원칙: Single Responsibility Principle): 작성된 클래스는 하나의 기능만 가지며 클래스가 제공하는 모든 서비스는 그 하나의 책임(변화의 축: axis of change)을 수행하는 데 집중되어 있어야 한다

플레이어의 상태를 나타내는 객체가 돌을 두는(place) 행위를 하는 것에 대해 어떻게 생각하시나요?

wondroid-world replied:
[ 관련 커밋 ] 85cf2c75fcd36f89eafb8c16cd735d359bc78ba8

[ 이유 ]
제이든! 제이든 말처럼 상태 패턴에서 행동을 한다는 게 맞지 않다는 사실을 깨달았습니다. 

제가 이해한 상태 패턴은 돌을 놓으면서 상태가 변한다에 초점을 두어야한다고 생각을 합니다. 그래서, 메서드 명은 state으로 변경을 했습니다. 

그리고 상태는 position만 받을 체, omokGame에게 시킴니다. 그리고 오목 게임에서 판단을 하고 그 결과에 따라, nextTurn이 다음 사람으로 넘겨주는 지, 비겼는 지, 아니면, 끝나는 지 넘겨주게 하였습니다.

KwonDae commented on src/main/kotlin/omok/model/rule/BlackWinRule.kt:
```diff
@@ -0,0 +1,23 @@
+            blink1 + blink2 == 0 && stone1 + stone2 == 4 -> true
+            else -> false
+        }
+    }
```
흑돌과 흰돌의 승리 기준이 다를까요?
OmokRule 에서 공통으로 사용할 수 있지 않을까요?

KwonDae commented on src/main/kotlin/omok/model/rule/BlackWinRule.kt:
```diff
@@ -0,0 +1,23 @@
+package omok.model.rule
+
+object BlackWinRule : OmokRule() {
```
클래스 이름은 흑돌이 이기는 룰인데
내부 함수는 checkWhiteWin 을 활용하고 있네요!

wondroid-world replied:
[ 관련 커밋 ] d3e89a4d8bb843f85deef35250396d83e2e8e389

아! 네이밍이 잘못되었네요! 수정했습니다!

KwonDae commented on src/main/kotlin/omok/model/player/state/BlackPlayerState.kt:
```diff
@@ -0,0 +1,39 @@
+        stoneState: StoneState,
+    ): PlayerState {
+        omokBoard.placeStone(position, stoneState)
+        val adaptedBoard = OmokAdapter.adaptOmokBoard(omokBoard)
+        val adaptedPoint = OmokAdapter.adaptOmokPoint(position)
+        return when {
+            BlackWinRule.validate(adaptedBoard, adaptedPoint) -> Win()
+            FourFourRule.validate(adaptedBoard, adaptedPoint) -> {
```
현재 State라는 객체가 OmokAdapter로 부터 위임받아 Rule을 검증하고 있네요.
OmokAdapter를 만들었으니, 어댑터 패턴을 먼저 적용해보고 상태패턴은 잠시 넣어두셔도 좋을 것 같습니다 🙂

KwonDae replied:
Rule을 활용해 현재 좌표를 검증하는건 OmokAdapter 의 역할이 아닐까요?
[어댑터 패턴](https://johngrib.github.io/wiki/pattern/adapter/)에 대해 한번 찾아보셔도 좋을 것 같아요

KwonDae commented on src/main/kotlin/omok/model/player/state/BlackPlayerState.kt:
```diff
@@ -0,0 +1,39 @@
+class BlackPlayerState(
+    private val count: Int = 0,
+) : PlayerState {
+    override fun isPlaceTurn(
+        omokBoard: OmokBoard,
+        position: Position,
+        stoneState: StoneState,
+    ): PlayerState {
+        omokBoard.placeStone(position, stoneState)
+        val adaptedBoard = OmokAdapter.adaptOmokBoard(omokBoard)
+        val adaptedPoint = OmokAdapter.adaptOmokPoint(position)
+        return when {
+            BlackWinRule.validate(adaptedBoard, adaptedPoint) -> Win()
+            FourFourRule.validate(adaptedBoard, adaptedPoint) -> {
+                omokBoard.doubleFour(position)
+                this
+            }
+
+            ThreeThreeRule.validate(adaptedBoard, adaptedPoint) -> {
+                omokBoard.doubleThree(position)
+                this
+            }
```
`현재 돌을 두려는 위치에 둘 수 있는가?` 는 누구의 역할일까요?

`Player`의 `put` 행위가 정의되어 있습니다. State를 거치지 않고서도 BlackPlayer가 돌을 두려 할 때 
Adapter를 통해 검증할 수 있는 흐름으로 개선해볼 수 있을까요?

wondroid-world replied:
[ 관련 커밋 ] 85cf2c75fcd36f89eafb8c16cd735d359bc78ba8

제이든 제가 상태 로직을 제대로 이야하지 못하고 짠거 같아서, 대대적인 수정을 진행했습니다.
상태에서는 돌을 두는 행동은 omokGame에게 메시징을 보내서, 행동을 시킵니다. 그리고 omokGame에게 결과를 달라고 메시징을 보냅니다. 그리고 결과에 따라서, 종료할지, 다음턴으로 넘길 지 수정을 했습니다.

KwonDae commented on src/main/kotlin/omok/model/rule/OmokAdapter.kt:
```diff
@@ -0,0 +1,22 @@
+import omok.model.stone.StoneState
+
+object OmokAdapter {
+    fun adaptOmokBoard(omokBoard: OmokBoard): List<List<Int>> {
+        val adapted = MutableList(omokBoard.ySize) { MutableList(omokBoard.xSize) { 0 } }
+        omokBoard.keys.forEach {
+            adapted[it.y.point - 1][it.x.point - 1] =
+                when (omokBoard.boardState(it)) {
+                    StoneState.BLACK -> 1
+                    StoneState.WHITE -> 2
+                    else -> 0
+                }
+        }
+        return adapted
+    }
```
위에서 코멘트로 남겨드린 어댑터패턴을 공부해보시면
OmokAdapter 객체내에서 Rule들을 어떻게 활용할 수 있을까요?

```kotlin
// OmokRuleAdapter
fun validPosition(..) {
  return rules.all { .. }
}
```

wondroid-world replied:
제이든 제가 이해한 바로는 어뎁터 패턴이 나의 코드를 다른 사람이 짠 코드에 맞출 때 정확하게 결합하기 위해서 사용한다고 알고 있습니다! 제이든이 말한바로는 어뎁터에서 도메인 로직까지 넣어서 처리하면 된다는 의도일까요??? 만약 그게 맞다면, 저의 생각에는 adpater가 너무 많은 것을 알고 있다고 생각이 들어요! 그리고 너무 많은 것을 처리한다는 생각이 드는 데, 제이든은 이부분은 허용을 해도 된다는 의도일까요??

(이부분은 제이든의 의견을 들어보고 수정을하고 싶어 아직 남겨두었습니다)

KwonDae reviewed (changes requested):
모찌 안녕하세요!
오목 미션 함께할 제이든입니다.
주로 객체들의 역할과 전체적인 구조에 대해 먼저 코멘트를 남겨드렸어요.
천천히 확인해주시고 어려운 부분은 편하게 DM주세요!

wondroid-world committed:
refactor: X 클래스 변경

- data class -> value class로 변경
- 생성자 이름 : point -> value로 변경

wondroid-world committed:
refactor: Y 클래스 변경

- data class -> value class로 변경
- test case 추가

wondroid-world committed:
refactor(Position): 부 생성자 삭제

wondroid-world committed:
docs: 기능 목록 추가

wondroid-world committed:
feat(Position): 위치에 대한 객체 구현
- 각 점은 초기에 돌이 놓여져 있지 않다.
- 돌을 위치에 놓으면 위치에는 돌이 있다.
- 자리에 돌이 있으면, 그 자리에 돌을 놓지 못한다.

wondroid-world committed:
refactor(Position): 주 생성자 private으로 변환 & 팩토리 생성자를 통한 생성만 가능

wondroid-world committed:
fix: 충돌 해결

wondroid-world committed:
chore: 필요없는 파일 삭제

wondroid-world committed:
feat: PlayerState 객체 구현

- PlayerState는 omokGame에서 결과만 받아서, 상태를 변화 시킴

wondroid-world committed:
refactor(State): nextTurn -> state로 변경

wondroid-world committed:
refactor: 오목판 객체 변경

- interface로 느슨한 결합 유도
- board는 불변 객체로 반환
- position 생성자 가시범위 변경

wondroid-world committed:
refactor: StoneState 중 Double_Three, Double_Four 삭제

wondroid-world committed:
chore: 파일 이동으로 인한 변경

wondroid-world committed:
refactor: 렌주룰 클래스 네이밍 변경

wondroid-world committed:
refactor: test 네이밍 수정

wondroid-world committed:
refactor: naming 수정

wondroid-world committed:
feat: stoneState 함수 추가

wondroid-world committed:
feat: game 인터페이스 구현

wondroid-world committed:
chore: 오목 판단 로직 삭제

wondroid-world committed:
refactor: 프로퍼티 접근 제어자 변경

wondroid-world committed:
feat: 렌즈룰 구현을 위한 필드 추가

wondroid-world committed:
feat: controller 구현

wondroid-world committed:
refactor: klint 수정

wondroid-world committed:
refactor: println 삭제

wondroid-world committed:
refactor: 메서드명 변경

wondroid-world commented:
안녕하세요, 제이든!

제이든 제가 아직 adapter 패턴에 대해서, 이해가 부족한것 같아서, 이번 수정은 상태 패턴에 초점을 두고 수정하게 되었습니다. adapter 패턴은 제이든의 의견을 듣고 수정하고 싶어서 남겨두었습니다. 

저번 블랙젝 수업때, 상태 패턴에 대해서 학습했는데,  배운 내용을 완전하게 학습하고 이해해보고 싶어서, 상태 패턴에 초점을 두고 수정하게되었습니다. 

그리고 제이든 말처럼 확장의 여지를 두고 개발을 해야한다고 생각해서, class 이전에 모두 interface로 두어서, 느슨한 결합을 하도록 노력했습니다.

KwonDae commented on src/main/kotlin/omok/model/omokGame/OmokGame.kt:
```diff
@@ -0,0 +1,21 @@
+import omok.model.board.Position
+import omok.model.stone.StoneState
+
+interface OmokGame {
```
> 그리고 제이든 말처럼 확장의 여지를 두고 개발을 해야한다고 생각해서, class 이전에 모두 interface로 두어서, 느슨한 결합을 하도록 노력했습니다.

OmokGame을 추상화 한 이유는 확장이었던거 같네요..!
오목 미션에서의 OmokGame은 추상화 대상일지 한번 고민해보시면 좋을거 같아요.
우리는 현재 어느 대상을 추상화 하면 좋을까요?
Stone? Player?

이전 미션들에서도 BlackjackGame, LottoGame 을 모두 추상화 했었을까요?

wondroid-world replied:
> 이전 미션들에서도 BlackjackGame, LottoGame 을 모두 추상화 했었을까요?

아니요! 처음에는 BackjackGame,, LottoGame을 모두 추상화하지 않았습니다. 하지만, 제가 interface랑 abstract class 차이를 배우면서, 제이슨한테 모든 class는 interface에서 출발했다라는 점을 배웠습니다. 우리 머리 속에서 interface로 시작했지만, 구체화되면서 class로 코드를 적는 것을 발견했습니다.

그래서, 이번에는 의식적으로 interface를 먼저 만들고 구현체를 써보자. 라는 마음에서, OmokGame을 추상화해서 적었습니다. 학습의 목적에서 위와같이 구현한 점 양해부탁드립니다.

> 우리는 현재 어느 대상을 추상화 하면 좋을까요?

제이든말처럼 굳이 추상화하지 않아도 될 대상이라고 생각도 동의합니다! 그럼 무엇을 추상화해야할까?를 고민했을 때, 확장의 여지가 있는 대상을 추상화해야겠다고 생각해요!

예로 오목에서 렌주룰이 아닌 다른 룰이 들어갈 수 있으니, rule을 추상화 해볼 수 있다고 생각해요. 

위와 비슷한 결로 추상화를 쓰는 이유는 결합도를 낮추기 위해서 사용을 한다고 생각해요.

KwonDae commented on src/main/kotlin/omok/model/rule/OmokAdapter.kt:
```diff
@@ -1,15 +1,15 @@
-            adapted[it.y.point - 1][it.x.point - 1] =
-                when (omokBoard.boardState(it)) {
+            adapted[it.y.value - 1][it.x.value - 1] =
+                when (omokBoard.stoneState(it)) {
```
https://github.com/woowacourse/kotlin-omok/pull/121#discussion_r2011119043

> adpater가 너무 많은 것을 알고 있다고 생각이 들어요! 그리고 너무 많은 것을 처리한다는 생각이 드는 데, 제이든은 이부분은 허용을 해도 된다는 의도일까요??

현재 OmokAdapter는 어떤 역할을 기대하고 있을까요?
[어댑터 패턴](https://ko.wikipedia.org/wiki/%EC%96%B4%EB%8C%91%ED%84%B0_%ED%8C%A8%ED%84%B4)에서의 어댑터는 `클래스의 인터페이스를 사용자가 기대하는 다른 인터페이스로 변환하는 패턴 ` 이라고 말하고 있습니다.

즉, OmokAdapter를 통해 렌주룰을 제외한 어느 룰이 추가되어도 Adapter를 통해 검증할 수 있어야 한다는 것이죠.
```kotlin
class OmokAdapter(rules: List<Rule>) {
   fun validPosition(..) {
      return rules.all { .. } 
   }
 ..
}
```

wondroid-world replied:
그럼, 렌주룰 말고 다른 라이브러리가 추가된다면, 동일한 adapter에서 처리를 하게 만들어 줘야한다는 의미일까요? 저는 처음 생각했을 때, 새로운 라이브러리가 추가 될때마다, 어뎁터를 추가해야한다는 생각을 가졌습니다. 하지만, 이렇게 생각하니, 확장할때마다 새로운 어뎁터 클래스를 만들어야된다는 단점이 있네요. 😥 

제이든의 말은 Rule interface 타입을 바로 구현하지말고, 인자로 받으라는 말인가요????

KwonDae commented on src/main/kotlin/omok/model/rule/FourFourRule.kt:
```diff
@@ -1,26 +1,25 @@
 package omok.model.rule
 
-object FourFourRule : OmokRule() {
+class FourFourRule : RenjuRule() {
```
3x4, 4x4 의 룰은 사실 렌주룰 이라는 하나의 룰 안에 종속되는 항목이 아닐까요?
그렇다면 렌주룰이 OmokRule을 구현하게 될 것 이고, 구현해주신 규칙들은 현재 미션에서 다루는 렌주룰 하위에 들어가는게 자연스럽지 않을까요?

KwonDae replied:
3x3, 4x4 "금수 규칙"은 오목의 렌주룰의 일부일거에요.
마치 축구게임에서 수많은 규칙 중 "라인아웃시 스로잉", "손에 닿으면 파울" 과 같이 여러 항목이 있는 것 처럼,
이에 비유하여 생각해보시면 어느 부분을 개선할 수 있을지 보일겁니다.

KwonDae commented on src/main/kotlin/omok/model/rule/OmokAdapter.kt:
```diff
@@ -1,15 +1,15 @@
 
-object OmokAdapter {
-    fun adaptOmokBoard(omokBoard: OmokBoard): List<List<Int>> {
+class OmokAdapter {
```
이름을 OmokRuleAdapter라고 불러야 더 이해하기 쉬워질 것 같다는 생각도 드네요

KwonDae commented on src/main/kotlin/omok/model/omokGame/OmokGameImpl.kt:
```diff
@@ -0,0 +1,59 @@
+        board.placeStone(position, stone)
+    }
+
+    override fun blackResult(
+        position: Position,
+        stoneState: StoneState,
+    ): TurnResult {
+        val adaptedBoard = adaptedBoard(board)
+        val adaptedPosition = adaptedPosition(position)
+        return when {
+            BlackWinRule().validate(adaptedBoard, adaptedPosition) -> TurnResult.WIN
+            FourFourRule().validate(adaptedBoard, adaptedPosition) -> throw IllegalStateException(ERROR_FOUR_FOUR)
+            ThreeThreeRule().validate(adaptedBoard, adaptedPosition) -> throw IllegalStateException(ERROR_THREE_THREE)
+            board.isFull() -> TurnResult.DRAW
+            else -> TurnResult.CONTINUE
+        }
+    }
```
OmokAdapter에서의 코멘트와 이어지는 내용입니다. 반드시 함께 고민해주세요.

현재 OmokGame에서 직접 Rule을 불러와(사실 렌주룰을 불러오는 것이 아닌, 렌주룰의 검증항목을 불러오고 있습니다)
그렇다면 이 구조에서 새로운 룰이 추가된다면 OmokGame에 직접 Rule을 호출해야 할까요?

이 부분에서 필요한건 사실 아래와 같은 흐름이 아닐까요?
```kotlin
val result = ruleAdapter.isValidPosition(board, position)
when (result) {
  .. -> WIN
 .. -> throw.. 
 .. -> CONTINUE
}
```

위와 같이 되어있다면 OmokGame 구현체가 직접 Rule이 추가되었는지, 어떤 룰을 적용해야 하는지 관심이 없어도 되겠네요!

wondroid-world replied:
```
override fun blackResult(
        position: Position,
        stoneState: StoneState,
    ): TurnResult {
        val omokRule = rule.validate(board, position, stoneState)
        return when (omokRule) {
            OmokRuleSet.FIVE_IN_A_ROW -> TurnResult.WIN
            OmokRuleSet.FOUR_FOUR -> throw IllegalStateException(ERROR_FOUR_FOUR)
            OmokRuleSet.THREE_THREE -> throw IllegalStateException(ERROR_THREE_THREE)
            OmokRuleSet.FULL -> TurnResult.DRAW
            else -> TurnResult.CONTINUE
        }
    }
```
제이든 제이든이 말한 의도가 이게 맞을까요?!!! 어뎁터 패턴을 이해하니, 굳이 변화하고 부르는 행동을 여기서 안해도 되어서, 코드가 조금 더 간결하게 변한거 같아요!

KwonDae commented on src/main/kotlin/omok/model/rule/BlackWinRule.kt:
```diff
@@ -1,23 +1,18 @@
 package omok.model.rule
 
-object BlackWinRule : OmokRule() {
+class BlackWinRule : RenjuRule() {
```
BlackWinRule 의 본질적인 역할에 대해 논의를 하고 싶어요.
렌주룰이 흑돌에게 해당하는 룰이 맞지만,
BlackWinRule은 렌주룰을 확장하는 관계일까요?

OmokGame에서 현재 Board와 Turn, Position 과 같은 상태를 알고 있으니
흑돌의 차례일 때 렌주룰을 적용하게끔 활용한다면 어떻게 될까요?

KwonDae commented on src/main/kotlin/omok/model/omokGame/OmokGameImpl.kt:
```diff
@@ -0,0 +1,59 @@
+        }
+    }
+
+    private fun adaptedBoard(board: Board): List<List<Int>> = OmokAdapter().adaptOmokBoard(board)
+
+    private fun adaptedPosition(position: Position): RulePosition = OmokAdapter().adaptOmokPoint(position)
```
https://github.com/woowacourse/kotlin-omok/pull/121#discussion_r2007105374

> 다른 룰이 추가되었을 때에도 사용할 수 있게 rule을 interface로 만들었습니다. 그리고 renjuRule이 Rule을 받아서 구현하도록 하였습니다. 이러면, 다른 룰이 추가되어도 omokGame에 바로 결합할 수 있는 형태가 된다고 생각합니다!

Rule이 추가된다면 OmokGame에 바로 결합되도록 해주셨군요..!
다만, OmokGame에서 직접 Rule을 통해 판단하는 것 보다, OmokGame 객체로부터 분리해보면 어떨까요?

지금 OmokAdapter에게 보다 주체적으로 역할을 부여할 수 있을거에요.

wondroid-world replied:
rule interface를 OmokAdapter가 구현했습니다. 그리고 오목 게임을 시작할 때, 이 omokAdapter를 받아서, 게임이 굴러가게 만들었는 데, 이러면 좀 더 omokAdapter가 주체적으로 일을 하는 걸까요?!!

wondroid-world replied:
[ 관련 커밋 ] ae84c5e5b0acb643badde1583bb993fb007d145c

KwonDae reviewed (changes requested):
피드백 반영하시느라 고생많으셨어요.
코드가 많이 개선되고 있는게 인상적이었습니다!

다만, 3,4 단계로 넘어가기 전 꼭 살펴보고 넘어갔으면 하는 부분들에 대해 코멘트를 남겨드렸어요.
시간이 많이 부족하기에 살펴보시고 다음 단계에서 함께 반영해주셔도 괜찮으니 
수요일 중으로 요청 주실 수 있는 만큼만 반영하시고 요청주세요.

어려운 부분은 편하게 DM 부탁드립니다!

wondroid-world committed:
refactor: 룰 라이브러리 변경

wondroid-world committed:
refactor: adapter 패턴 변경

wondroid-world committed:
feat: stoneState 오버로딩

wondroid-world committed:
refactor: 파라미터 타입 변경

wondroid-world committed:
feat: 룰 주입

wondroid-world committed:
fix: 오목 룰 판정할 수 있도록 수정

KwonDae commented on src/main/kotlin/omok/model/rule/Rule.kt:
```diff
@@ -1,8 +1,14 @@
+        board: Board,
+        position: Position,
+        stoneState: StoneState,
+    ): OmokRuleSet
```
https://github.com/woowacourse/kotlin-omok/pull/121#discussion_r2013286731

추상화를 하여 결합도를 낮추는 시도 좋았습니다 👍 
다만, 수업시간에 다뤘던 내용처럼 `라이브러리` 를 사용하는 것 처럼 구현해야 한다면
Rule 인터페이스는 특정 도메인에 의존성을 갖으면 안될거에요.

현재의 omok.model 에 의존성을 갖게 하면 어떤 일이 발생할까요?

wondroid-world replied:
제이든의 말처럼, rule interface의 함수인 validate 파라미터로 오목보드를 받는 것조차 결합도를 높이는 일이군요! 만약에 이렇게 결합도가 높으면, 다른 체스 게임이나, 바둑 등에 재사용하고 싶어도 하지 못할 거 같네요!

차라리 interface를 지우고 오목 게임은 오목 어뎁터만 알 수 있도록 바꾸어봤습니다!

wondroid-world replied:
[ 관련 커밋 ] 477ba0f12618bab2821f310cf0ddae1e59e53d70

KwonDae commented on src/main/kotlin/omok/model/rule/OmokRuleAdapter.kt:
```diff
@@ -0,0 +1,67 @@
+import omok.model.omokGame.OmokRuleSet
+import omok.model.stone.StoneState
+
+class OmokRuleAdapter : Rule {
+    override fun validate(
+        board: Board,
+        position: Position,
+        stoneState: StoneState,
+    ): OmokRuleSet =
+        when (stoneState) {
+            StoneState.BLACK -> blackStoneState(board, position)
+            StoneState.WHITE -> whiteStoneState(board, position)
+            else -> throw IllegalStateException("이 상황에서는 값을 판단할 수 없습니다.")
+        }
+
+    private fun blackStoneState(
+        board: Board,
+        position: Position,
+    ): OmokRuleSet {
+        val adaptedBoard: List<List<Int>> = board.toMatrix()
+        val adaptedPosition: Pair<Int, Int> = position.toCoordinates()
+        return when {
+            BlackWinRule.validate(adaptedBoard, adaptedPosition) -> OmokRuleSet.FIVE_IN_A_ROW
+            FourFourRule.validate(adaptedBoard, adaptedPosition) -> OmokRuleSet.FOUR_FOUR
+            ThreeThreeRule.validate(adaptedBoard, adaptedPosition) -> OmokRuleSet.THREE_THREE
+            else -> OmokRuleSet.CONTINUE
+        }
+    }
+
+    private fun whiteStoneState(
+        board: Board,
+        position: Position,
+    ): OmokRuleSet {
+        val adaptedBoard: List<List<Int>> = board.toMatrix()
+        val adaptedPosition: Pair<Int, Int> = position.toCoordinates()
+        if (WhiteWinRule.validate(adaptedBoard, adaptedPosition)) {
+            return OmokRuleSet.FIVE_IN_A_ROW
+        }
+        return OmokRuleSet.CONTINUE
+    }
```
좋습니다 👍 
이제 Adapter는 `돌을 두었을 때 해당 좌표에 둘 수 있는 돌인지?` 에 대해 검증 역할을 갖게 되었어요.

https://github.com/woowacourse/kotlin-omok/pull/121#discussion_r2013444758

> 그럼, 렌주룰 말고 다른 라이브러리가 추가된다면, 동일한 adapter에서 처리를 하게 만들어 줘야한다는 의미일까요?

이것이 어댑터의 역할이 아닐까요?

- 다른 라이브러리가 추가되는 것
- Rule interface 타입을 바로 구현하지말고, 인자로 받는 것

은 별도로 생각해야 할 포인트 입니다.
Rule 인터페이스는 Adapter의 역할(~룰 인진 관심이 없고, OmokRule의 행동을 명시한다)
Adapter는 하나 이상의 Rule에 대해 사용할 수 있도록 연결 역할을 해준다.

지금 모찌가 구현한 Rule 인터페이스는 Adapter의 행동을(네이밍때문에 헷갈릴 수 있겠네요)
OmokRule 인터페이스는 오목의 룰에 대해 추상화를 해주신게 아닐까요?

KwonDae replied:
사실 지금의 어댑터는 OmokRuleAdapter보다
OmokRenjuRuleAdapter 에 가깝겠네요

wondroid-world replied:
맞아요! 처음에 각각의 라이브러리마다 어뎁터가 있어야 된다고 생각했지만, 수업을 듣고 크루들이랑 이야기하면서, adapter는 모든 library와 domain을 이어주는 controller와 같다고 생각이 들었습니다. 이제야 조금 더 adapter의 기능을 알게 되었습니다!

KwonDae commented on src/main/kotlin/omok/model/omokGame/OmokGameImpl.kt:
```diff
@@ -2,35 +2,30 @@ package omok.model.omokGame
-            FourFourRule().validate(adaptedBoard, adaptedPosition) -> throw IllegalStateException(ERROR_FOUR_FOUR)
-            ThreeThreeRule().validate(adaptedBoard, adaptedPosition) -> throw IllegalStateException(ERROR_THREE_THREE)
-            board.isFull() -> TurnResult.DRAW
+        val omokRule = rule.validate(board, position, stoneState)
+        return when (omokRule) {
+            OmokRuleSet.FIVE_IN_A_ROW -> TurnResult.WIN
+            OmokRuleSet.FOUR_FOUR -> throw IllegalStateException(ERROR_FOUR_FOUR)
+            OmokRuleSet.THREE_THREE -> throw IllegalStateException(ERROR_THREE_THREE)
+            OmokRuleSet.FULL -> TurnResult.DRAW
```
조금 더 개선할 수 있다면
지금의 OmokGameImpl은 렌주룰에 강하게 의존하고 있는게 아닐까요?

4x4 -> throw
3x3 -> throw

와 같은 로직이 이곳에 위치한다면 만약 룰이 변경되었을 때 어댑터에서만 룰을 갈아끼우는게 아닌
도메인 로직을 수정해야 할거에요.

Adapter 에서 승부를 판단하여 해당 돌이 우승인지만 결과를 반환한다면 어떨까요?

wondroid-world replied:
아! 제이든 말은 omokGame에서는 굳이 룰을 해석하지 않고 게임이 진행될 수 있는 상태인지만 확인하면 된다는 이야기죠! 제이든의 말처럼 어뎁터에서 룰의 결과가 아닌 턴의 결과를 반환할 수 있도록 변환했습니다.

``` OmokRuleAdapter
    private fun blackStoneState(
        board: Board,
        position: Position,
    ): TurnResult {
        val adaptedBoard: List<List<Int>> = board.toMatrix()
        val adaptedPosition: Pair<Int, Int> = position.toCoordinates()
        return when {
            BlackWinRule.validate(adaptedBoard, adaptedPosition) -> TurnResult.WIN
            FourFourRule.validate(adaptedBoard, adaptedPosition) -> throw IllegalStateException(
                ERROR_FOUR_FOUR
            )
            ThreeThreeRule.validate(adaptedBoard, adaptedPosition) ->  throw IllegalStateException(
                ERROR_THREE_THREE
            )
            else -> TurnResult.CONTINUE
        }
    }

```

[ 관련 커밋 ] d822c2795aa194f00c275f2112b4eff745eb4314

KwonDae commented on src/main/kotlin/omok/library/WhiteWinRule.kt:
```diff
@@ -0,0 +1,23 @@
+package omok.library
+
+object WhiteWinRule : OmokRule(WHITE_STONE, BLACK_STONE) {
```
BlackWinRule/WhiteWinRule 도 서로 로직이 크게 달라보이지 않아요
두 돌의 승리 기준이 분리되어야 할까요?

wondroid-world replied:
제이든! 저의 생각에는 승리 기준이 분리 되어야한다고 생각합니다!

BlackWinRule과 WhiteWinRule은 코드 구조상 굉장히 유사합니다. 하지만, 이 둘은 돌마다 다른 승리 조건을 갖고 있어서 역할을 분리하는 게 더 좋다고 생각합니다! 흑돌은 엄격한 기준을 가지고 있지만, 하얀돌은 오목만 되면 된다는 조건을 가지고 있습니다. 만약 나중에 렌주룰이 바뀐다면 blackWinRule만 수정하면 되고 하얀돌 로직에는 영향이 없도록 할 수 있을 거 같아요!

KwonDae reviewed (approved):
1,2 단계 피드백 반영하시느라 고생많으셨어요.
3,4 단계 진행하시면서 도움이 되었으면 하는 부분에 코멘트를 남겨드렸습니다.

시간이 부족하시겠지만 진행에 참고 해주시고,
어려운 부분은 편하게 DM주셔도 됩니다 :) 
마지막까지 파이팅이에요!
